### PR TITLE
[back/front] Allow to make some entity/fields mandatory

### LINKED ISSUE #1850: Allow to make some entity/fields mandatory
## Use case

Allow to make some entity/fields mandatory. For example always have a TLP marking is required.

## Current Workaround

Tell users not to forget the fields considered required.

## Proposed Solution

Be able to customize mandatory fields by entity type in the settings of the platform.

## Additional Information

As there can be several things in the markings, it would be even better to be able to choose which type of marking is made mandatory by the entity.

Ideally it should be possible to do this for several types of entities in order to have a capitalization model that is always the same for the organization.

## If the feature request is approved, would you be willing to submit a PR?

### PR DESCRIPTION
**Pull request by RomuDeuxfois**
### Proposed changes

*
*

### Related issues

* [issues 1850](https://github.com/OpenCTI-Platform/opencti/issues/1850)

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: opencti-platform/opencti-front/src/private/components/analysis/groupings/GroupingCreation.js [hunk 1]
```diff
@@ -91,12 +92,18 @@ const groupingMutation = graphql`
...
-});
+const groupingValidation = (t) => {
+  let shape = {
+    name: Yup.string().required(t('This field is required')),
+    confidence: Yup.number(),
+    context: Yup.string().required(t('This field is required')),
+    description: Yup.string().nullable(),
+  };
+
+  shape = useCustomYup('Grouping', shape, t);
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/private/components/analysis/groupings/GroupingCreation.js]: Not a big fan of this notation with a let and a rewrite.
I would prefer a simple
const basicShape = {....}
const shape = useCustomYup('Grouping', basicShape , t);
useCustomYup should also directly generate the shape i think

File: opencti-platform/opencti-front/src/utils/hooks/useEntitySettings.ts [hunk 2]
```diff
@@ -35,4 +41,48 @@ export const useIsEnforceReference = (id: string): boolean => {
...
+    .find((node) => id === node.target_type && node.attributes_configuration !== null);
+
+  if (!entitySetting || !entitySetting.attributes_configuration) {
+    return null;
+  }
+
+  return JSON.parse(entitySetting.attributes_configuration);
+};
+
+export const useCustomYup = <TNextShape extends ObjectShape>(id: string, existingShape: TNextShape, t: (message: string) => string): TNextShape => {
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/utils/hooks/useEntitySettings.ts]: Doest really makes sens to have t in parameter.
Must be removed and use the correct hook directly inside.
const { t } = useFormatter();

File: opencti-platform/opencti-graphql/src/database/middleware-utils.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
+import Ajv from 'ajv';
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/database/middleware-utils.ts]: File to rename to schema-validator.ts? to discuss

File: opencti-platform/opencti-graphql/src/database/middleware.js [hunk 4]
```diff
@@ -2387,49 +2370,18 @@ const upsertElementRaw = async (context, user, element, type, updatePatch) => {
...
-      }
-    }
-  }
-  // -- Upsert single refs specific for cyber observable
-  // TODO JRI Upsert single refs specific for cyber observable
-  // endregion
   // region generic elements
   // -- Upsert multiple refs for other stix elements
-  const metaInputFields = STIX_META_RELATIONSHIPS_INPUTS;
+  let metaInputFields;
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/database/middleware.js:2385]: Use a ternary?

File: opencti-platform/opencti-graphql/src/database/middleware.js [hunk 5]
```diff
@@ -2387,49 +2370,18 @@ const upsertElementRaw = async (context, user, element, type, updatePatch) => {
...
+  if (isStixCyberObservable(type)) {
+    metaInputFields = stixCyberObservableFieldsForType(type);
+  } else {
+    metaInputFields = stixNotCyberObservableFields();
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/database/middleware.js]: Strange name function stixNotCyberObservableFields

File: opencti-platform/opencti-graphql/src/database/middleware.js [hunk 6]
```diff
@@ -2866,8 +2819,7 @@ const buildEntityData = async (context, user, input, type, opts = {}) => {
     R.assoc('creator_id', user.internal_id),
     R.dissoc('update'),
     R.dissoc('file'),
-    R.omit(STIX_META_RELATIONSHIPS_INPUTS),
-    R.omit(STIX_CYBER_OBSERVABLE_RELATIONSHIPS_INPUTS),
+    R.omit(schemaRelationsRefDefinition.getName()),
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/database/middleware.js]: To discuss, not sure to understand how 2 remove can be converted to only one

File: opencti-platform/opencti-graphql/src/database/stix.ts [hunk 7]
```diff
@@ -1297,20 +1281,8 @@ export const checkStixCyberObservableRelationshipMapping = (fromType: string, to
   return R.includes(relationshipType, targetRelations);
 };
 
-export const CHECK_META_RELATIONSHIP_VALUES: { [k: string]: (fromType: string, toType: string) => boolean } = {
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/database/stix.ts:1300]: Simple marker to talk about how it works.

File: opencti-platform/opencti-graphql/src/domain/stixDomainObject.js [hunk 8]
```diff
@@ -253,8 +252,7 @@ export const stixDomainObjectEditField = async (context, user, stixObjectId, inp
   }
   // Check is a real update was done
   const updateWithoutMeta = R.pipe(
-    R.omit(STIX_META_RELATIONSHIPS_INPUTS),
-    R.omit(STIX_CYBER_OBSERVABLE_RELATIONSHIPS_INPUTS),
+    R.omit(schemaRelationsRefDefinition.getName()),
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/domain/stixDomainObject.js]: Same commment about the single omit

File: opencti-platform/opencti-graphql/src/modules/administrativeArea/administrativeArea.ts [hunk 9]
```diff
@@ -33,13 +34,13 @@ const ADMINISTRATIVE_AREA_DEFINITION: ModuleDefinition<StoreEntityAdministrative
...
-    { name: 'longitude', type: 'string', multiple: false, upsert: true },
-    { name: 'x_opencti_location_type', type: 'string', multiple: false, upsert: false },
-    { name: 'x_opencti_stix_ids', type: 'string', multiple: true, upsert: true },
-    { name: 'x_opencti_workflow_id', type: 'string', multiple: false, upsert: true },
+    { name: 'name', type: 'string', mandatoryType: 'external', multiple: false, upsert: true },
+    { name: 'description', type: 'string', mandatoryType: 'customizable', multiple: false, upsert: true },
+    { name: 'latitude', type: 'string', mandatoryType: 'customizable', multiple: false, upsert: true },
+    { name: 'longitude', type: 'string', mandatoryType: 'customizable', multiple: false, upsert: true },
+    entityLocationType,
+    xOpenctiStixIds,
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/modules/administrativeArea/administrativeArea.ts]: Required?  Looks like not defined for other entity

File: opencti-platform/opencti-graphql/src/modules/grouping/grouping.ts [hunk 10]
```diff
@@ -29,10 +28,10 @@ const GROUPING_DEFINITION: ModuleDefinition<StoreEntityGrouping, StixGrouping> =
     },
   },
   attributes: [
-    { name: 'name', type: 'string', multiple: false, upsert: true },
-    { name: 'description', type: 'string', multiple: false, upsert: true },
-    { name: 'context', type: 'string', multiple: false, upsert: true },
-    { name: 'x_opencti_graph_data', type: 'string', multiple: false, upsert: false },
+    { name: 'name', type: 'string', mandatoryType: 'external', multiple: false, upsert: true },
+    { name: 'description', type: 'string', mandatoryType: 'customizable', multiple: false, upsert: true },
+    { name: 'context', type: 'string', mandatoryType: 'external', multiple: false, upsert: true },
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/modules/grouping/grouping.ts:33]: vocabulary mandatory must be computed from vocabulary-types.ts

File: opencti-platform/opencti-graphql/src/schema/stixCyberObservableRelationship.ts [hunk 11]
```diff
@@ -62,234 +92,291 @@ export const RELATION_VALUES = 'values';
...
+};
+const values: RelationRefDefinition = { // Not in standard
+  inputName: INPUT_VALUES,
+  databaseName: RELATION_VALUES,
+  stixName: 'values_refs',
+  mandatoryType: 'no',
+  multiple: true,
+};
+// Not used ? TODO: remove
+// const xOpenctiLinkedTo: RelationRefDefinition = { // Not in standard
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/schema/stixCyberObservableRelationship.ts]: To discuss about relastionships commented in this file

File: opencti-platform/opencti-graphql/src/schema/stixMetaRelationship.ts [hunk 12]
```diff
@@ -22,98 +78,185 @@ export const RELATION_KILL_CHAIN_PHASE = 'kill-chain-phase'; // kill_chain_phase
...
-R.map(
-  (stixInternalMetaRelationshipType) => schemaTypes.registerAttributes(stixInternalMetaRelationshipType, stixMetaRelationshipsAttributes),
-  STIX_INTERNAL_META_RELATIONSHIPS
+
+schemaRelationsRefDefinition.registerRelationsRef(ABSTRACT_STIX_DOMAIN_OBJECT, [createdBy, objectMarking, objectLabel, externalReferences]);
+schemaRelationsRefDefinition.registerRelationsRef(ABSTRACT_STIX_CYBER_OBSERVABLE, [objectMarking]);
+schemaRelationsRefDefinition.registerRelationsRef(ABSTRACT_STIX_RELATIONSHIP, [
+  {
+    ...createdBy,
+    mandatoryType: 'no',
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/schema/stixMetaRelationship.ts]: Why forcing mandatory to no here?

**Review by richard-julien — CHANGES_REQUESTED** [review]

**Comment by RomuDeuxfois** [comment]
Bug found : 
- enforce ref enable -> button update but form not valid - FIXED
- enforce ref enable -> update form and crash
- enforce ref enable -> create a new ref and it's direclty add
- adminstrativArea -> author not show required
- adminstrativArea -> mandatory field, field patch empty field work - FIXED

**Pull request merged**